### ISSUE TITLE
Failing tests (suspected to be related to changes in deal.II)

### ISSUE THREAD

**Issue by nfehn** [exadg/exadg#241]
```
The following tests FAILED:
	  2 - tests/euler_vortex.release (Failed)
	  7 - tests/convective_explicit_time_int.release (Failed)
	  9 - tests/explicit_time_int.release (Failed)
```

Since I updated today to the dealii/master branch and did now see failing tests before, I guess that these failing tests must have been introduced by recent changes in deal.II.

**Comment by nfehn** [comment]
@kronbichler @peterrum The failing tests occur for compressible Navier-Stokes, where we did not have recent changes in ExaDG if I remember correctly. Since you are the dealii experts, could you go through the changes in dealii related to MatrixFree/FEEvaluation and point out potentially problematic PRs?

**Comment by nfehn** [comment]
see comment by @kronbichler on this topic here: https://github.com/exadg/exadg/pull/236#issuecomment-1175993053

**Comment by nfehn** [comment]
@kronbichler Do you suggest that we shift the parameter `DEGREE_MAX` from ExaDG to deal.II? This would be unfortunate from a user perspective in my opinion.

**Comment by kronbichler** [comment]
The parameter `DEGREE_MAX` (or equivalent deal.II name) is present in deal.II already, it is just that we choose to define it differently here. From the perspective of ExaDG development, the code would simplify, because we can skip the extra instantiations. As you say, it does shift the responsibility to the user for picking an alternative degree in case deal.II's maximum of 6 is not enough, but this should only be a performance problem and not a correctness problem. (Unless something else did break in the slow paths of deal.II in the meantime.)

**Comment by kronbichler** [comment]
For me, the test `tests/convective_explicit_time_int.release`, which picks degree 15, passes, whereas the other two, picking degree 7, fail. The interesting fact is that degree 6 works for those two tests (ok, apart from different convergence rates of course.) Given that I compile exadg with a different `DEGREE_MAX` than the official setting, this indicates that the code mixes the template instantiations of deal.II with those of ExaDG, making the getting out of sync.

**Comment by kronbichler** [comment]
> Which value should one choose, 2, 5, 7, 15? I think one could also formulate it as a design goal that the compilation parameters of a general library like dealii should not depend on projects build on top of deal.II.

I see the point. To expand on why things are like that, deal.II also profits from this infrastructure in certain places (like the evaluation of the MappingQ, some project functions, etc). All those use cases had manual solutions or slower code before they could also use the infrastructure we had created for ExaDG. So from that point of view, it appeared as a win-win case.

**Comment by kronbichler** [comment]
Compiling ExaDG with `DEGREE_MAX=6`, in sync with deal.II's settings, and allowing `dealii::Patterns::Integer(1,30)` in https://github.com/exadg/exadg/blob/707bd29887e82aae585db36ba14ded8ce1027217/include/exadg/utilities/resolution_parameters.h#L54 and https://github.com/exadg/exadg/blob/707bd29887e82aae585db36ba14ded8ce1027217/include/exadg/utilities/resolution_parameters.h#L59 fixes all tests. Could this be that splitting the compilation into the face factory and regular factory with the RT code added?

**Comment by kronbichler** [comment]
More precisely, I believe the problem is that `internal::FastEvaluationSupported` sits in the file going to the `FEEvaluation` class (cell integrals), which the `FEFaceEvaluation` object is also using. I think the problem would be fixed by adding a `fast_evaluation_supported` function to `FEFaceEvaluationFactory`, which would internally do the same thing. That would then redirect all symbol lookup into the same compilation unit of the class, and work around these problems. But I am not sure how robust it would be, as technically the symbol lookup details depend on the operating system and on the linker.

**Comment by kronbichler** [comment]
The quick fix for ExaDG would be to set `DEGREE_MAX=6` in line with deal.II, which would then always result in a robust method. The question is if
- we want to maintain a different setting for the pre-compiled versions of the polynomial degree without touching the installed/built variant of deal.II (otherwise, one can pass that compile option)
- in case we want the independent setting from deal.II, how we make sure in a robust way that symbols in different libaries, `libdeal_II.so` and `libexadg.so`, do not get mixed up. Static linkage would be an option, but that is clearly not the solution.

If the suggested way to set the choice of fast path is through the deal.II library, we should work towards making the option in deal.II a proper one - right now, we just pass a define flag to all compilation units of deal.II that then picks up the correct macro in the right file, which is not guaranteed to keep working.

**Comment by kronbichler** [comment]
> I fear they will stop using ExaDG just because of the difficult handling

I don't think setting one additional option in the compilation of deal.II make a big difference in the grant scheme of thing. We also suggest to use an appropriate `-march=` option, to use a number of third-party libraries from deal.II, some of them as required dependency, others as optional, and so on. If we want to lower the threshold for users to get started, I would not worry about this additional parameter (to give users more performance), but about providing a spack recipe or similar all-inclusive solutions that build exadg with deal.II and all other dependencies out-of-the-box.

**Comment by nfehn** [comment]
@kronbichler How do you explain that tests are failing in ExaDG? I still did not understand this. Also, I believe some tests in ExaDG use degrees higher than 6? What does this imply.

**Comment by nfehn** [comment]
ok, sound like we should be working towards passing this parameter to the cmake configuration of deal.II ... and not using or introducing this parameter at all in ExaDG.

Why do you suggest `dealii::Patterns::Integer(1,30)`, is there a deal.II-internal limit of 30? If not, those checks do not really help and we can remove them. However, I would insist that deal.II prints a warning about slow execution also in release mode!

**Comment by kronbichler** [comment]
To support the hypthesis that I expressed in previous posts, I tested with adding
```
  template <int dim, typename Number, typename VectorizedArrayType>
  bool
  FEFaceEvaluationGatherFactory<dim, Number, VectorizedArrayType>::
    fast_evaluation_supported(const unsigned int given_degree,
                              const unsigned int n_q_points_1d)
  {
    return instantiation_helper_run<1, FastEvaluationSupported>(
      given_degree, n_q_points_1d);
  }
```
to deal.II, and can confirm that this fixes the failing tests also when deal.II and ExaDG have different settings for the maximal compiled degree. We could roll in this feature, but I am not sure if that solution is robust, as I merely create another symbol among hundreds of other symbols in the `FEFaceEvaluationGatherFactory` surrounding, and what gets picked is not clear. For example, I cannot say whether this code actually picks up ExaDG's pre-compiled polynomial degrees or the one from deal.II, without looking at a trace of the execution. (I previously checked that we pick the code from ExaDG, but now I'm not so sure any more. I might not be the most qualified person to make this kind of constructions as I do not know much about how libraries are actually composed or how the run time picks a specific implementation.)

**Comment by kronbichler** [comment]
> Why do you suggest dealii::Patterns::Integer(1,30), is there a deal.II-internal limit of 30? If not, those checks do not really help and we can remove them.

There is no technical limitation, I just hardcoded it to something for my internal tests, I did not want to suggest a solution. We could still check that the number is non-negative (or do we have to exclude 0?), but that's it.

> However, I would insist that deal.II prints a warning about slow execution also in release mode!

This is something we need to do in ExaDG when we parse the parameters for the polynomial degree, deal.II does not have the means to do it: Where would that check to be done? Once we set up `FEEvaluation`/`FEFaceEvaluation`? How to ensure that it only prints it once and only on one MPI rank (`FEEvaluation` does not know about the MPI universe)? How do we ensure that we do not pay a run-time cost every time we use `FEEvaluation`, as that is a performance-sensitive class? What if the user is fine with the slow path, as long as the work gets done - how to prevent printing the warning at that spot? It's theoretically possible to do all this, but ExaDG is the right place to print the warning because we print extensive parameter lists already and because we know that the wrong choice hurts our programs.

**Comment by nfehn** [comment]
I do not see any use in constraining DEGREE_MAX in ExaDG to the one in deal.II to have a robust behavior. Since removing DEGREE_MAX from deal.II does not seem to be a debatable option, it seems clear what we need to do, i.e. remove it from ExaDG?

**Comment by nfehn** [comment]
> >    However, I would insist that deal.II prints a warning about slow execution also in release mode!

> This is something we need to do in ExaDG when we parse the parameters for the polynomial degree, deal.II does not have the means to do it: Where would that check to be done? Once we set up FEEvaluation/FEFaceEvaluation? How to ensure that it only prints it once and only on one MPI rank (FEEvaluation does not know about the MPI universe)? How do we ensure that we do not pay a run-time cost every time we use FEEvaluation, as that is a performance-sensitive class? What if the user is fine with the slow path, as long as the work gets done - how to prevent printing the warning at that spot? It's theoretically possible to do all this, but ExaDG is the right place to print the warning because we print extensive parameter lists already and because we know that the wrong choice hurts our programs.

I thought @peterrum suggested something like this, but maybe I misunderstood this. Still, I think it is more involved, because there are / will be many other projects like ExaDG. Do we want that all have to take care of this? Another argument is that we then connect deal.II and ExaDG more strongly, accessing deal.II's DEGREE_MAX in ExaDG. From my perspective, this is not a single-responsibility solution. Again, I would then raise the question why not doing all instantiations in ExaDG (for the sake of single-responsibility), but we had this discussion.

**Comment by kronbichler** [comment]
> Since removing DEGREE_MAX from deal.II does not seem to be a debatable option,

I would be happy to debate things, but it is not me unwilling to debate it here. We cannot remove the parameter there because that parameter is **the mechanism** to control what degrees get instantiate in the respective `.templates.h` file, which both deal.II and ExaDG rely on. I had hoped to be able to bypass deal.II's pre-compiled code if we compile these `.templates.h` files again with a different setting of that parameter, making the symbols reliably come from the library that comes earlier in the link line (libexadg). The problem is that we cannot chose a different name, because we have to know it inside the deal.II header `fe_evaluation.h` to know what function to call. The only way to get independent would be to by-pass all `evaluate`/`integrate`/`gather_evaluate`/`integrate_scatter` calls and re-implement in exadg. Maybe 500-800 lines of code, and the need to adapt to changes in deal.II as we would need to rely on deal.II's internal interfaces.

**Comment by nfehn** [comment]
Another thing is that if we print this warning in ExaDG, tests will fail or not depending on which configuration one has set in deal.II, because the parameter or the warning is printed. Then, we could de-activate such warnings for tests. But I acutally don't like such solutions that become quite complicated.

**Comment by nfehn** [comment]
We could assert that we never use a degree higher than deal.II's DEGREE_MAX in ExaDG. If this condition is not fulfilled, tests and other simulations will fail in ExaDG and the user gets the message "You used degree > DEGREE_MAX. Make sure to adjust DEGREE_MAX in the configuration of deal.II in order to use the specified degree.". 

This is more restrictive but maybe clearer or more transparent to use??

**Comment by nfehn** [comment]
We have to be aware that users will use ExaDG in publications and publish performance numbers, but they have not been aware of this parameter. It is hard to criticize this if we cannot provide a user-friendly solution.

**Comment by nfehn** [comment]
My suggestion would be that @kronbichler or @peterrum provide a patch for how to adjust deal.II's and ExaDG's config-script and remove the DEGREE_MAX parameter from ExaDG. I can start from that point to find a solution for ExaDG and doing the necessary checks, output, etc.

**Comment by kronbichler** [comment]
To follow up on my post https://github.com/exadg/exadg/issues/241#issuecomment-1176201572, i.e., this observation here:

> I tested with adding to deal.II, and can confirm that this fixes the failing tests also when deal.II and ExaDG have different settings for the maximal compiled degree.

I think we should also work towards making the infrastructure in deal.II more robust. It won't change the need #244 in my opinion because we still can't be sure if the dynamic library loader picks one version or the other, so we would still not be 100% sure to control the performance side in ExaDG, but at least we should avoid getting runtime errors. Let me open an issue in deal.II so we do not forget the reasoning here.

**Comment by nfehn** [comment]
this issue should have been fixed by #244 , so I will close it
